src/common.jl

```diff
@@ -30,3 +30,16 @@ struct Identity end
 \(::Identity, x) = copy(x)
 ldiv!(::Identity, x) = x
 ldiv!(y, ::Identity, x) = copyto!(y, x)
+
+"""
+Conjugated and unconjugated dot products
+"""
+abstract type DotType end
+struct ConjugatedDot <: DotType end
+struct UnconjugatedDot <: DotType end
+
+_norm(x, ::ConjugatedDot) = norm(x)
+_dot(x, y, ::ConjugatedDot) = dot(x, y)
+
+_norm(x, ::UnconjugatedDot) = sqrt(sum(xₖ^2 for xₖ in x))
+_dot(x, y, ::UnconjugatedDot) = sum(prod, zip(x, y))
```

Comment:
`sqrt(sum(xₖ -> xₖ^2, x))` for the norm, so that it can use pairwise summation (iterators are not indexable)?

Reply:
Wow, this change makes a significant difference in performance:
```
julia> x = rand(1000);

julia> @btime sum(xₖ^2 for xₖ in $x);
  1.221 μs (0 allocations: 0 bytes)

julia> @btime sum(xₖ->xₖ^2, $x);
  85.675 ns (0 allocations: 0 bytes)
```
But it doesn't seem like indexibility is the only condition for the pairwise summation to kick in.  I tried to achieve a similar enhancement by pairwise summation in `_dot(x, y, ::UnconjugatedDot) = sum(prod, zip(x, y))` by replacing `zip(x,y)` with an indexible custom type, but it didn't work as expected.  Looked into the code in detail, and it seems that `mapreduce_impl` will need to be implemented for the custom type replacing `zip(x,y)`...